File: src/cli_tool/caw_cli.cpp
#include <gflags/gflags.h>

#include <iostream>
#include <string>

#include "../caw/client/cpp_client/caw_client.h"

DEFINE_string(registeruser, "", "Registers the given username");
DEFINE_string(user, "", "Logs in as the given username");
DEFINE_string(caw, "", "Creates a new Caw with the given text");
DEFINE_int64(reply, -1,
             "Indicates that the new caw is a reply to the given id");
DEFINE_string(follow, "", "Starts following the given username");
DEFINE_int64(read, -1, "Reads the caw thread starting at the given id");
DEFINE_bool(profile, false,
            "Gets the user’s profile of following and followers");
DEFINE_string(stream, "", "Streams all new caws containing the given hashtag");
DEFINE_string(hook, "", "Hooks the given function");
DEFINE_string(unhook, "", "Unhooks the given function");
DEFINE_bool(hookall, false, "Hooks all stored functions");
DEFINE_bool(unhookall, false, "Unhooks all stored functions");

int main(int argc, char *argv[]) {
  // Checking and handeling arguments
  if (argc < 2) {
    gflags::ShowUsageWithFlags(argv[0]);
    return 1;
  }

  gflags::ParseCommandLineFlags(&argc, &argv, true);

  // Ensuring a username is always specified
  // With exception of registeruser
  if (FLAGS_registeruser == "" && FLAGS_user == "" && FLAGS_hook == "" &&
      FLAGS_unhook == "" && FLAGS_unhookall == false && FLAGS_stream == "" &&
      FLAGS_hookall == false) {
    std::cerr << "You cannot perform any action other than hooking/unhooking "
                 "without specifying a username"
              << std::endl;
    std::cerr << "Please consider using the --user <username> flag"
              << std::endl;
    return 1;
  }

  if (FLAGS_hookall && FLAGS_unhookall) {
    std::cerr << "You cannot hook all and unhook all at the same time"
              << std::endl;
    return 1;
  }

  if (FLAGS_reply != -1 && FLAGS_caw == "") {
    std::cerr << "You cannot reply without specifying a caw" << std::endl;
    std::cerr << "Please consider using the --caw <caw text> flag" << std::endl;
    return 1;
  }

  // Error checking done, handeling arguments

  // Creating a client instance
  CawClient client;

  if (FLAGS_registeruser != "") {
    bool registerSuccess = client.RegisterUser(FLAGS_registeruser);
  } else if (FLAGS_reply != -1) {
    bool replySuccess = client.PostCaw(FLAGS_user, FLAGS_caw, FLAGS_reply);
  } else if (FLAGS_caw != "") {
    // Since we have already checked for reply
    // We know this would be a new caw and not a reply
    // According to specs, we specify a new caw using -1
    bool postSuccess = client.PostCaw(FLAGS_user, FLAGS_caw, -1);
  } else if (FLAGS_follow != "") {
    bool followSuccess = client.FollowUser(FLAGS_user, FLAGS_follow);
  } else if (FLAGS_read != -1) {
    bool readSuccess = client.ReadCaw(FLAGS_read);
  } else if (FLAGS_profile) {
    bool profileSuccess = client.GetProfile(FLAGS_user);
  } else if (FLAGS_stream != "") {
    bool streamSuccess = client.StreamHashtag(FLAGS_stream);
  } else if (FLAGS_hook != "") {
    bool hookSuccess = client.HookFunction(FLAGS_hook);
  } else if (FLAGS_unhook != "") {
    bool unhookSuccess = client.UnhookFunction(FLAGS_unhook);
  } else if (FLAGS_hookall) {
    bool hookallSuccess = client.HookAll();
  } else if (FLAGS_unhookall) {
    bool unhookallSuccess = client.UnhookAll();
  }

  return 0;
}

File: src/cli_tool/caw_cli.go
package main

import (
	"flag"
	"fmt"
	"os"

	caw "github.com/floan/csci499_floan/src/caw/client/go_client"
)

// Defining command line flags
var (
	registeruserPtr = flag.String("registeruser", "", "Registers the given username")
	userPtr         = flag.String("user", "", "Logs in as the given username")
	cawPtr          = flag.String("caw", "", "Creates a new Caw with the given text")
	replyPtr        = flag.Int("reply", -1,
														 "Indicates that the new caw is a reply to the given id")
	followPtr  = flag.String("follow", "", "Starts following the given username")
	readPtr    = flag.Int("read", -1, "Reads the caw thread starting at the given id")
	profilePtr = flag.Bool("profile", false,
												 "Gets the user’s profile of following and followers")
	hookPtr      = flag.String("hook", "", "Hooks the given function")
	unhookPtr    = flag.String("unhook", "", "Unhooks the given function")
	hookallPtr   = flag.Bool("hookall", false, "Hooks all stored functions")
	unhookallPtr = flag.Bool("unhookall", false, "Unhooks all stored functions")
)

func main() {
	// Checking and handling arguments
	if len(os.Args) < 2 {
		fmt.Println("You have not used any command line arguments. Please try again.")
		flag.PrintDefaults()
		os.Exit(1)
	}

	flag.Parse()

	// Ensuring a username is always specified
	// With exception of registeruser
	if *registeruserPtr == "" && *userPtr == "" && *hookPtr == "" &&
		*unhookPtr == "" && *unhookallPtr == false &&
		*hookallPtr == false {
		err := fmt.Errorf("You cannot perform any action other than " +
											"hooking/unhooking without specifying a username. \n" +
											"Please consider using the --user <username> flag.")
		fmt.Println(err.Error())
		os.Exit(1)
	}

	if *hookallPtr && *unhookallPtr {
		err := fmt.Errorf("You cannot hook all and unhook all simultaneously.")
		fmt.Println(err.Error())
		os.Exit(1)
	}

	if *replyPtr != -1 && *cawPtr == "" {
		err := fmt.Errorf("You cannot reply without specifying a caw. \n" +
											"Please consider using the --caw <caw text> flag.")
		fmt.Println(err.Error())
		os.Exit(1)
	}

	// Error checking done, now
	// we can handle arguments

	// Creating client instance
	client := caw.CreateCawClient()

	if *registeruserPtr != "" {
		client.RegisterUser(*registeruserPtr)
	} else if *replyPtr != -1 {
		client.PostCaw(*userPtr, *cawPtr, *replyPtr)
	} else if *cawPtr != "" {
		// Since we have already checked for reply
		// We know this would be a caw and not a reply
		// According to specs, we specify a new caw using -1
		client.PostCaw(*userPtr, *cawPtr, -1)
	} else if *followPtr != "" {
		client.FollowUser(*userPtr, *followPtr)
	} else if *readPtr != -1 {
		client.ReadCaw(*readPtr)
	} else if *profilePtr {
		client.GetProfile(*userPtr)
	} else if *hookPtr != "" {
		client.HookFunction(*hookPtr)
	} else if *unhookPtr != "" {
		client.UnhookFunction(*unhookPtr)
	} else if *hookallPtr {
		client.HookAll()
	} else if *unhookallPtr {
		client.UnhookAll()
	}
}
